### PR TITLE
Fix/charts features

### PR DESCRIPTION
**Pull request by ShrawanBK**

### CODE REVIEW

File: src/components/ChartModal/ScatterChartConfig/index.tsx [hunk 1]
```diff
@@ -25,20 +26,63 @@ interface Props<T> {
...
+        index,
+        value,
+        onValueChange,
+        options,
+    } = props;
+
+    // FIXME: memoize
+    const keySelector = (item: NumericOption<T>) => item.key;
+    const labelSelector = (item: NumericOption<T>) => item.title;
+    const groupSelector = (item: NumericOption<T>) => item.category;
```
Comment (tnagorra) [inline src/components/ChartModal/ScatterChartConfig/index.tsx]: Need to memoize this

File: src/components/ChartModal/ScatterChartConfig/index.tsx [hunk 2]
```diff
@@ -52,16 +96,53 @@ function ScatterChartConfig<T>(props: Props<T>) {
...
+                if (mappedData.length === 1) {
+                    return [
+                        ...mappedData,
+                        defaultData[0],
+                    ];
+                }
+                const [firstDataValue, secondDataValue] = mappedData;
+                return [firstDataValue, secondDataValue];
+            }
+            return defaultData;
```
Comment (tnagorra) [inline src/components/ChartModal/ScatterChartConfig/index.tsx]: Shortcircuit if-else

File: src/components/ChartModal/ScatterChartConfig/index.tsx [hunk 3]
```diff
@@ -52,16 +96,53 @@ function ScatterChartConfig<T>(props: Props<T>) {
...
+                }
+                const mappedData = editableData.map((e) => {
+                    const opt = options.find(o => o.key === e.key);
+                    if (!opt) {
+                        return undefined;
+                    }
+                    return {
+                        id: randomString(),
+                        optionName: opt.key,
+                    };
```
Comment (tnagorra) [inline src/components/ChartModal/ScatterChartConfig/index.tsx]: Use ternary operators

File: src/components/DendogramTree/index.tsx [hunk 4]
```diff
@@ -1,40 +1,25 @@
...
+    onHideDendrogram?: (id: string | undefined) => void
 }
 
-let NODE_WIDTH = 164;
+let NODE_WIDTH = 260;
 const NODE_HEIGHT = 24;
 const NODE_CIRCLE_RADIUS = 4;
 const NODE_GAP_Y = 12;
-let NODE_GAP_X = 96;
+let NODE_GAP_X = 120;
```
Comment (tnagorra) [inline src/components/DendogramTree/index.tsx]: Why aren't these constants?

File: src/components/DropdownMenu/index.tsx [hunk 5]
```diff
@@ -50,6 +52,7 @@ function DropdownMenu(props: Props) {
                 elementRef={buttonRef}
                 onClick={handleShowDropdown}
                 disabled={disabled}
+                data-tut={dataTut}
```
Comment (tnagorra) [inline src/components/DropdownMenu/index.tsx:55]: What is `data-tut`?

File: src/components/IndicatorMap/index.tsx [hunk 6]
```diff
@@ -154,24 +191,107 @@ function IndicatorMap(props: Props) {
...
-        const numberedBbox = bbox.split(',').map(b => Number(b));
+        const numberedBbox = bbox.split(',').map(b => +b);
         if (numberedBbox.length === 4) {
             return numberedBbox as Bound;
         }
-        return undefined;
-    }, [regionListResponse?.results]);
+        return defaultBounds;
+    }, [regionListResponse?.results, isMinimized]);
+    const selectedRegionState: MapStateItem[] | undefined = useMemo(() => {
```
Comment (tnagorra) [inline src/components/IndicatorMap/index.tsx]: This is never undefined

File: src/components/IndicatorMap/index.tsx [hunk 7]
```diff
@@ -154,24 +191,107 @@ function IndicatorMap(props: Props) {
...
+        if (!selectedRegionId) {
+            return [];
+        }
+        return [{
+            id: selectedRegionId,
+            value: 1,
+        }];
+    }, [selectedRegionId]);
+
+    const hoveredRegionState: MapStateItem[] | undefined = useMemo(() => {
```
Comment (tnagorra) [inline src/components/IndicatorMap/index.tsx]: This is never undefined

File: src/components/PolyChart/BiAxialChart/index.tsx [hunk 8]
```diff
@@ -125,35 +123,45 @@ export function BiAxialChartUnit<T extends object>(props: BiAxialChartUnitProps<
...
-    const hasLongTitles = averageLength > 5;
+    const hasLongTitles = useMemo(
+        () => {
+            if (longTilesShown) {
+                return true;
+            }
+            return averageLength > 5; // acceptableLength = 5
+        },
+        [longTilesShown, averageLength],
+    );
```
Comment (tnagorra) [inline src/components/PolyChart/BiAxialChart/index.tsx]: No need to memoize this.
We can write this in single expression

File: src/components/PolyChart/ScatterChart/index.tsx [hunk 9]
```diff
@@ -78,29 +67,17 @@ export function ScatterChartUnit<T extends object>(props: ScatterChartUnitProps<
...
-                .slice(0, limit.count);
-        },
-        [data, limit, valueSelector],
-    );
+    const [firstData, secondData] = useMemo(() => {
+        if (!scatterData) {
+            return [undefined, undefined];
+        }
+        return scatterData;
+    }, [scatterData]);
```
Comment (tnagorra) [inline src/components/PolyChart/ScatterChart/index.tsx]: No need to memoize this.
We can write this in single expression

File: src/components/ProgramSelector/index.tsx [hunk 10]
```diff
@@ -177,7 +202,45 @@ function ProgramSelector(props: Props) {
         selectedProgram.filter(item => item.startsWith('program')).map(item => Number(item.split('-')[1])),
     );
 
+    const selectedComponentOriginal = unique(
+        selectedProgram.filter(item => item.startsWith('subprogram')).map(item => item.replace('subprogram-', '')),
+    );
```
Comment (tnagorra) [inline src/components/ProgramSelector/index.tsx]: Need to memoize this calculation

File: src/components/ProgramSelector/index.tsx [hunk 11]
```diff
@@ -177,7 +202,45 @@ function ProgramSelector(props: Props) {
...
+    const rawSubMarkerOptions = subMarkerListResponse?.results;
+    const rawProgramOptionsWithoutComponentSectors = programListResponse?.results;
+    const rawComponentOptions = componentListResponse?.results;
+    const rawProgramOptions = useMemo(() => {
+        if (!rawComponentOptions) {
+            return undefined;
+        }
+        if (!rawProgramOptionsWithoutComponentSectors) {
+            return undefined;
+        }
```
Comment (tnagorra) [inline src/components/ProgramSelector/index.tsx]: We can use `||` to merge both of the if statements

File: src/schema/index.ts [hunk 12]
```diff
@@ -1,7 +1,7 @@
 import Dict, { basicTypes, Schema } from '@togglecorp/ravl';
 import { isProduction } from '#config/env';
 
-/* eslint-disable @typescript-eslint/camelcase */
+/* eslint-disable camelcase */
```
Comment (tnagorra) [inline src/schema/index.ts:4]: Any reason why this was changed?

File: src/views/About/TermsAndConditionsPage/LeftTitleItem/index.tsx [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    } = props;
+
+    const hash = useHash();
+
+    const isSelectedItem: boolean = useMemo(() => {
+        if (!hash) {
+            return false;
+        }
+        return id === +hash;
+    }, [id, hash]);
```
Comment (tnagorra) [inline src/views/About/TermsAndConditionsPage/LeftTitleItem/index.tsx]: No need to memoize this.
We can write this in single expression

File: src/views/About/TermsAndConditionsPage/TermAndConditionItem/index.tsx [hunk 14]
```diff
@@ -19,14 +22,34 @@ export default function TermAndConditionItem(props: TermAndConditionItemProps) {
...
     } = props;
 
+    const hash = useHash();
+
+    const isSelectedItem: boolean = useMemo(() => {
+        if (!hash) {
+            return false;
+        }
+        return tc.id === +hash;
+    }, [tc.id, hash]);
```
Comment (tnagorra) [inline src/views/About/TermsAndConditionsPage/TermAndConditionItem/index.tsx]: No need to memoize this.
We can write this in single expression

File: src/views/Dashboard/ProgramDetails/index.tsx [hunk 15]
```diff
@@ -38,6 +38,9 @@ function ProgramDetails(props: Props) {
     } = useContext<DomainContextProps>(DomainContext);
 
     const [selectedTab, setSelectedTab] = useState<TabOptionKeys>('charts');
+    const programIdList = unique(
+        programs.filter(item => item.startsWith('program')).map(item => Number(item.split('-')[1])),
+    );
```
Comment (tnagorra) [inline src/views/Dashboard/ProgramDetails/index.tsx]: Memoize please!

File: src/views/Dashboard/DateRangeSelector/index.tsx [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+            const convStartDate = new Date(defaultStartDate).getFullYear();
+            const convEndDate = new Date(defaultEndDate).getFullYear();
+            return [convStartDate, convEndDate];
+        },
+        [defaultStartDate, defaultEndDate],
+    );
+    const selectedSliderDateRange = useMemo(() => {
+        const convStartDate = new Date(defaultStartDate).getFullYear();
+        const convEndDate = new Date(defaultEndDate).getFullYear();
+        const defYear = [convStartDate, convEndDate];
```
Comment (tnagorra) [inline src/views/Dashboard/DateRangeSelector/index.tsx]: Why not just read defaultYears?

File: src/views/Dashboard/DateRangeSelector/index.tsx [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+        if (startDate) {
+            defYear[0] = new Date(startDate).getFullYear();
+            if (endDate) {
+                defYear[0] = new Date(startDate).getFullYear();
```
Comment (tnagorra) [inline src/views/Dashboard/DateRangeSelector/index.tsx]: Why assign the same value again?

File: src/views/Dashboard/Tooltip/index.tsx [hunk 18]
```diff
@@ -240,6 +287,26 @@ const Tooltip = (props: Props) => {
...
+                            {program?.sector?.length > 0 && (
+                                <h4 className={styles.sectorsHeader}>
+                                    Sectors
+                                </h4>
+                            )}
+                            <div className={styles.sectorList}>
+                                {(unique(
+                                    program.sector,
+                                    d => d.sector,
+                                )?.map(sector => (
```
Comment (tnagorra) [inline src/views/Dashboard/Tooltip/index.tsx:305]: We may need to move this outside of the jsx block

File: src/views/RegionProfile/index.tsx [hunk 19]
```diff
@@ -154,30 +170,28 @@ function RegionProfile(props: Props) {
...
-        [regionProfileResponse?.activeSectors],
-    );
+    const activeSectors: ActiveSectors[] | undefined = regionProfileResponse?.activeSectors;
+
+    const activeSectorsForChart: ProfileChartData[] | undefined = activeSectors?.map(sect => ({
+        name: sect.name,
+        id: sect.id,
+        value: sect.value,
+        key: sect.key,
+    }));
```
Comment (tnagorra) [inline src/views/RegionProfile/index.tsx:197]: any reason why we are mapping this?

File: src/views/RegionProfile/index.tsx [hunk 20]
```diff
@@ -223,14 +220,205 @@ function RegionProfile(props: Props) {
...
+                f => !hiddenDendrogramNames.includes(String(f.name)),
+            );
+        },
+        [hiddenDendrogramNames, mappedDendogramData],
+    );
+
+    const resetIndicatorsShown = useMemo(
+        () => {
+            const totalHiddenFiveWData = !!hiddenFiveWDataKeys?.length;
+            const totalHiddenIndicators = !!hiddenIndicatorsIds?.length;
```
Comment (tnagorra) [inline src/views/RegionProfile/index.tsx]: It's better to check explicitly for zero and undefined

File: src/views/RegionProfile/index.tsx [hunk 21]
```diff
@@ -342,30 +553,71 @@ function RegionProfile(props: Props) {
...
+                                            )}
+                                        >
+                                            <IoMdClose
+                                                className={styles.icon}
+                                            />
+                                        </Button>
+                                    </div>
                                 </div>
-                                {mappedDendogramData.map(res => (
+                                {filteredDendrogramData && filteredDendrogramData.map(res => (
```
Comment (tnagorra) [inline src/views/RegionProfile/index.tsx]:
```suggestion
                                {filteredDendrogramData?.map(res => (
```

File: src/views/ProgramProfile/Charts/index.tsx [hunk 22]
```diff
@@ -294,47 +179,55 @@ function InfographicsCharts(props: Props) {
...
+    //         [...chartSettings
+    //             .map(item => item.dependencies)
+    //             .filter(isDefined)
+    //             .flat(),
+    //         ].filter(i => !!indicatorMapping[i]),
+    //         item => item,
+    //     ).sort(),
+    //     [chartSettings, indicatorMapping],
+    // );
+
```
Comment (tnagorra) [inline src/views/ProgramProfile/Charts/index.tsx]: Let's remove this!

File: src/views/ProgramProfile/Charts/index.tsx [hunk 23]
```diff
@@ -294,47 +179,55 @@ function InfographicsCharts(props: Props) {
...
     );
 
+    const filteredFiveWData = extendedFiveWList
+        .filter(data => data.allocatedBudget > 0)
+        .sort((foo, bar) => compareNumber(
+            foo.allocatedBudget,
+            bar.allocatedBudget,
+            1,
+        ))
+        .slice(0, 10);
```
Comment (tnagorra) [inline src/views/ProgramProfile/Charts/index.tsx]: Let's memoize this

File: src/views/ProgramProfile/Charts/index.tsx [hunk 24]
```diff
@@ -369,6 +262,32 @@ function InfographicsCharts(props: Props) {
...
+        [setExpandableDefaultChart],
+    );
+
+    const expandableDefaultChartSettings = useMemo(
+        () => {
+            const chartSetting = defaultChartSettings.find(c => c.id === expandableDefaultChart);
+            if (!chartSetting) {
+                return undefined;
+            }
+            return chartSetting;
```
Comment (tnagorra) [inline src/views/ProgramProfile/Charts/index.tsx]: Do we even need this!

File: src/views/RegionProfile/NumberOutput/index.tsx [hunk 25]
```diff
@@ -1,24 +1,45 @@
...
+        onHideData,
+        id,
+    } = props;
+
+    const handleHideData = useCallback(
+        () => {
+            if (!onHideData) {
+                return;
+            }
+            onHideData(id);
```
Comment (tnagorra) [inline src/views/RegionProfile/NumberOutput/index.tsx]: We don't need to shortcircuit for this usecase

File: src/views/RegionProfile/Charts/index.tsx [hunk 26]
```diff
@@ -389,36 +243,81 @@ function InfographicsCharts(props: Props) {
         [chartSettings, expandableChart],
     );
 
+    const expandableDefaultChartSettings = useMemo(
+        () => {
+            const chartSetting = defaultChartSettings.find(c => c.id === expandableDefaultChart);
+            if (!chartSetting) {
+                return undefined;
+            }
+            return chartSetting;
```
Comment (tnagorra) [inline src/views/RegionProfile/Charts/index.tsx]: Why do this?

**Review by tnagorra — APPROVED** [review]
Need to fix some minor changes!

**Review by tnagorra — CHANGES_REQUESTED** [review]
Need to finish minor changes

File: src/Root/styles.css [hunk 27]
```diff
@@ -0,0 +1,475 @@
...
+        .recharts-default-tooltip {
+            white-space: unset!important;
+
+            p {
+                max-width: 300px;
+            }
+        }
+    }
+
+    .rc-slider {
```
Comment (AdityaKhatri) [inline src/Root/styles.css]: Move rc-slider related classnames to the component where rc slider is used

File: src/components/ChartModal/ScatterChartConfig/index.tsx [hunk 28]
```diff
@@ -52,16 +96,53 @@ function ScatterChartConfig<T>(props: Props<T>) {
...
+                },
+                {
+                    id: randomString(),
+                },
+            ];
+            if (editableChartData) {
+                const editableData = editableChartData.data;
+                if (!editableData) {
+                    return defaultData;
+                }
```
Comment (AdityaKhatri) [inline src/components/ChartModal/ScatterChartConfig/index.tsx]: Move this logic outside to remove one level of nesting.

File: src/components/DendogramTree/index.tsx [hunk 29]
```diff
@@ -73,36 +57,66 @@ function DendogramTree(props: DendogramTreeInterface) {
...
     };
 
     const childCount = secondLevelChildrens + thirdLevelChildrens;
 
+    const handleHideDendrogram = useCallback(
+        () => {
+            if (!onHideDendrogram) {
+                return;
+            }
+            onHideDendrogram(treeData.name);
```
Comment (tnagorra) [inline src/components/DendogramTree/index.tsx]:
```suggestion
            if (onHideDendrogram) {
                onHideDendrogram(treeData.name);
            }
```

File: src/components/PolyChart/BarChart/index.tsx [hunk 30]
```diff
@@ -127,31 +133,37 @@ export function BarChartUnit<T extends object>(props: BarChartUnitProps<T>) {
...
-                    actions.style.display = 'flex';
-                }
+    const xCompWidth = useMemo(
+        () => {
+            if (layout === 'horizontal') {
+                return undefined;
             }
+            return hasLongTitles ? 140 : 86;
+        },
+        [layout],
```
Comment (tnagorra) [inline src/components/PolyChart/BarChart/index.tsx]:
```suggestion
        [layout, hasLongTiles],
```

File: src/utils/downloadChart.ts [hunk 31]
```diff
@@ -0,0 +1,27 @@
...
+        const actions = divRef.current
+            .getElementsByClassName(className)[0] as HTMLDivElement;
+        if (actions) {
+            actions.style.display = 'none';
+        }
+        const png = await html2canvas(
+            divRef.current, {
+                scale: 3,
+            },
+        ).then(canvas => canvas.toDataURL('image/png', 1.0));
```
Comment (tnagorra) [inline src/utils/downloadChart.ts]: Why mix usage of then on async await

File: src/views/Dashboard/index.tsx [hunk 32]
```diff
@@ -219,6 +236,16 @@ const Dashboard = (props: Props) => {
...
     const [toolTipMinimized, , , toggleToolTipMinimized] = useBasicToggle();
     const [filterButtonHidden, hideFilterButton, showFilterButton] = useBasicToggle();
+    const [selectedStatus, setSelectedStatus] = useState<StatusTabOptionKeys>('all');
+
+    const onSelectStatus = useCallback(
+        (statusTabKey: StatusTabOptionKeys) => {
+            setSelectedStatus(statusTabKey);
+        },
+        [setSelectedStatus],
+    );
```
Comment (tnagorra) [inline src/views/Dashboard/index.tsx]: Why use useCallback?

File: src/views/RegionProfile/Indicators/index.tsx [hunk 33]
```diff
@@ -43,37 +52,35 @@ export default function Indicators(props: IndicatorProps) {
...
+        resettableIndicatorCategories,
     } = props;
 
+    const groupedIndicatorsData = useMemo(() => listToGroupList(
+        indicatorsData?.sort(
+            (a, b) => compareString(
+                a.indicator,
+                b.indicator,
+            ),
+        ),
```
Comment (tnagorra) [inline src/views/RegionProfile/Indicators/index.tsx]: Should not call sort on array as it mutates it. Should create a copy first!

File: src/views/RegionProfile/Indicators/GroupedIndicator/index.tsx [hunk 34]
```diff
@@ -0,0 +1,124 @@
...
+        () => resettableIndicatorCategories?.includes(category),
+        [resettableIndicatorCategories, category],
+    );
+
+    const onHideIndicatorId = useCallback(
+        (id: string) => {
+            if (!handleHideIndicatorId) {
+                return;
+            }
+            handleHideIndicatorId(category, id);
```
Comment (tnagorra) [inline src/views/RegionProfile/Indicators/GroupedIndicator/index.tsx]: No need to shortcircuit this

File: src/views/RegionProfile/Indicators/GroupedIndicator/index.tsx [hunk 35]
```diff
@@ -0,0 +1,124 @@
...
+        },
+        [category, onHideCategory],
+    );
+
+    const handleResetCategory = useCallback(
+        () => {
+            if (!onResetCategory) {
+                return;
+            }
+            onResetCategory(category);
```
Comment (tnagorra) [inline src/views/RegionProfile/Indicators/GroupedIndicator/index.tsx]: No need to circuit this

File: src/views/RegionProfile/Indicators/GroupedIndicator/index.tsx [hunk 36]
```diff
@@ -0,0 +1,124 @@
...
+        onResetCategory,
+        resettableIndicatorCategories,
+    } = props;
+
+    const handleHideCategory = useCallback(
+        () => {
+            if (!onHideCategory) {
+                return;
+            }
+            onHideCategory(category);
```
Comment (tnagorra) [inline src/views/RegionProfile/Indicators/GroupedIndicator/index.tsx]: No need to shortcircuit this

File: src/components/DropdownMenu/index.tsx (hunk 5 above)
Comment (tnagorra) [reply src/components/DropdownMenu/index.tsx:55]: Shrawan: It's for walkthrough

**Pull request merged**